platforms/jvm/testing-jvm/src/integTest/groovy/org/gradle/testing/NonClassBasedTestingIntegrationTest.groovy

```diff
@@ -0,0 +1,498 @@
+
+            testing.suites.test {
+                ${setupSuiteWithEngineFixture()}
+
+                targets.all {
+                    testTask.configure {
+                        scanForTestDefinitions = true
+                    }
+                }
+            }
+        """
+
+        when:
+        fails("test", "-S")
+
+        then:
+        if (OperatingSystem.current().isWindows()) {
+            sourcesPresentAndNoTestsFound()
+        } else {
+            failureCauseContains("Cannot access input property 'candidateDefinitionDirs' of task ':test'. Accessing unreadable inputs or outputs is not supported.")
+            failureCauseContains("Failed to create MD5 hash for file: ${badFile.absolutePath} (Permission denied)")
+        }
+
+        cleanup:
+        // restore read permission for cleanup
+        badFile.setReadable(true, false)
+    }
+
+    def "resource-based test engine detects and executes test definitions (excluding jupiter engine = #excludingJupiter)"() {
+        given:
+        buildFile << """
+            plugins {
+                id 'java-library'
+            }
+
+            ${mavenCentralRepository()}
+
+            testing.suites.test {
+                ${setupSuiteWithEngineFixture()}
+
+                targets.all {
+                    testTask.configure {
+                        scanForTestDefinitions = true
+
+                        options {
+                            includeEngines("rbt-engine")
+                            if ($excludingJupiter) {
+                                excludeEngines("junit-jupiter")
+                            }
+                        }
+                    }
+                }
+            }
+        """
+
+        writeTestDefinitions()
+
+        when:
+        succeeds("test", "--info")
+
+        then:
+        nonClassBasedTestsExecuted()
+
+        where:
+        excludingJupiter << [true, false]
+    }
+
+    def "resource-based test engine detects and executes test definitions in custom location"() {
+        String customLocation = "src/test/some-other-place"
+
+        given:
+        buildFile << """
+            plugins {
+                id 'java-library'
+            }
+
+            ${mavenCentralRepository()}
+
+            testing.suites.test {
+                ${setupSuiteWithEngineFixture()}
+
+                targets.all {
+                    testTask.configure {
+                        scanForTestDefinitions = true
+                        testDefinitionDirs.setFrom(project.layout.projectDirectory.file("$customLocation"))
+
+                        options {
+                            includeEngines("rbt-engine")
+                        }
+                    }
+                }
+            }
+        """
+
+        writeTestDefinitions(customLocation)
+
+        when:
+        succeeds("test", "--info")
+
+        then:
+        nonClassBasedTestsExecuted()
+    }
+
+    def "empty custom test definitions location skips"() {
```

Comment:
I think this is a duplicate test if we no longer have a default test definition dir.